prov/gni/include/gnix_tags.h

```diff
@@ -0,0 +1,416 @@
+ * @return -FI_EINVAL, if any invalid parameters were given
+ *         FI_SUCCESS, otherwise
+ */
+static inline int _gnix_posted_tag_storage_init(
+		struct gnix_tag_storage *ts,
+		struct gnix_tag_storage_attr *attr)
+{
+	return _gnix_tag_storage_init(ts, attr, _gnix_match_posted_tag);
+}
+
+/**
+ * @brief initialization function for unexpected tag storages
+ *
+ * @param ts          tag storage pointer
+ * @param attr        tag storage attributes
+ * @param match_func  match function to be used on individual list elements
+ * @return -FI_EINVAL, if any invalid parameters were given
+ *         FI_SUCCESS, otherwise
+ */
+static inline int _gnix_unexpected_tag_storage_init(
+		struct gnix_tag_storage *ts,
+		struct gnix_tag_storage_attr *attr)
+{
+	return _gnix_tag_storage_init(ts, attr, _gnix_match_unexpected_tag);
+}
+
+/**
+ * @brief destroys a tag storage and releases any held memory
+ *
+ * @param ts
+ * @return -FI_EINVAL, if the tag storage is in a bad state
+ *         -FI_EAGAIN, if there are tags remaining in the tag storage
+ *         FI_SUCCESS, otherwise
+ */
+int _gnix_tag_storage_destroy(struct gnix_tag_storage *ts);
+
+/**
+ * @brief inserts a gnix_fab_req into the tag storage
+ *
+ * @param ts           pointer to the tag storage
+ * @param tag          tag associated with fab request
+ * @param req          gnix fabric request
+ * @param ignore       bits to ignore in tag (only applies to posted)
+ * @param addr_ignore  bits to ignore in addr (only applies to posted)
+ * @return
+ *
+ * @note if ts is a posted tag storage, 'req->ignore_bits' will be set to
+ *         the value of 'ignore'.
+ *
+ * @note if ts is a posted tag storage and ts->attr.use_src_addr_matching
+ *         is enabled, 'req->addr_ignore_bits' will be set to the value
+ *         of 'addr_ignore'.
+ */
+int _gnix_insert_by_tag(
+		struct gnix_tag_storage *ts,
+		uint64_t tag,
+		struct gnix_fab_req *req,
+		uint64_t ignore,
+		uint64_t addr_ignore);
+
+/**
+ * @brief removes a request from the tag storage by tag and address
+ *
+ * @param ts           pointer to the tag storage
+ * @param tag          tag to remove
+ * @param ignore       bits to ignore in tag
+ * @param flags        fi_tagged flags
+ * @param context      fi_context associated with tag
+ * @param addr         gnix_address associated with tag
+ * @param addr_ignore  bits to ignore in address
+ * @return NULL, if no entry found that matches parameters
+ *         otherwise, a non-null value pointing to a gnix_fab_req
+ *
+ * @note ignore parameter is not used for posted tag storages
+ * @note addr_ignore parameter is not used for posted tag storages
+ */
+struct gnix_fab_req *_gnix_remove_by_tag_and_addr(
+		struct gnix_tag_storage *ts,
+		uint64_t tag,
+		uint64_t ignore,
+		uint64_t flags,
+		void *context,
+		struct gnix_address *addr,
+		uint64_t addr_ignore);
+
+/**
+ * @see _gnix_remove_by_tag_and_addr
+ */
+static inline struct gnix_fab_req *_gnix_remove_by_tag(
```

Comment:
A single matching interface might work for all cases:

_gnix_match_by_tag(..., flags, ...)

Flags values of FI_CLAIM, FI_DISCARD, FI_PEEK could control the behavior.  addr value of FI_ADDR_UNSPEC wold match any address.